magnum-zio/src/main/scala/com/augustnagro/magnum/magzio/Transactor.scala

```diff
@@ -0,0 +1,200 @@
+package com.augustnagro.magnum.magzio
+
+import com.augustnagro.magnum.{DbCon, DbTx, SqlLogger}
+import zio.{IO, Semaphore, Task, Trace, UIO, ZIO}
+
+import java.sql.Connection
+import javax.sql.DataSource
+import scala.reflect.ClassTag
+import scala.util.control.{ControlThrowable, NonFatal}
+
+/** Transactor lets you customize a transaction or connection's behavior. It is
+  * a parameter to the [[transact]] and [[connect]] methods.
+  */
+class Transactor private (
+    /** Datasource to be used */
+    private val dataSource: DataSource,
+    /** Logging configuration */
+    private val sqlLogger: SqlLogger,
+    /** Customize the underlying JDBC Connections */
+    private val connectionConfig: Connection => Unit,
+    /** Number of threads in your connection pool. This helps magzio be more
+      * memory efficient by limiting the number of blocking pool threads used.
+      * Not needed if using the ZIO virtual-thread based blocking executor
+      */
+    private val semaphore: Option[Semaphore]
+):
+
+  def withSqlLogger(sqlLogger: SqlLogger): Transactor =
+    new Transactor(
+      dataSource,
+      sqlLogger,
+      connectionConfig,
+      semaphore
+    )
+
+  def withConnectionConfig(connectionConfig: Connection => Unit): Transactor =
+    new Transactor(
+      dataSource,
+      sqlLogger,
+      connectionConfig,
+      semaphore
+    )
+
+  /** Executes a given query on a given Transactor, returning UIO.
+    *
+    * Re-implementation for ZIO of
+    * [[com.augustnagro.magnum.connect(dataSource: DataSource)]]
+    *
+    * Usage:
+    * {{{
+    *   import com.augustnagro.magnum.magzio.*
+    *
+    *   connectUIO(transactor):
+    *      repo.findById(id)
+    * }}}
+    */
+  def connectUIO[A](q: DbCon ?=> A)(using Trace): UIO[A] = connectTask(q).orDie
+
+  /** Executes a given query on a given Transactor, returning IO.
+    *
+    * Re-implementation for ZIO of
+    * [[com.augustnagro.magnum.connect(dataSource: DataSource)]]
+    *
+    * Usage:
+    * {{{
+    *   import com.augustnagro.magnum.magzio.*
+    *
+    *   connectIO(transactor):
+    *      repo.findById(id)
+    * }}}
+    */
+  def connectIO[E: ClassTag, A](q: DbCon ?=> A)(using Trace): IO[E, A] =
```

Comment:
This `ClassTag` might become annoying for the user at the use-site. Why do we need it? 🤔